mbus-api/src/receive.rs

```diff
@@ -0,0 +1,166 @@
+use super::*;
+
+/// Type safe wrapper over a message bus message which decodes the raw
+/// message into the actual payload `S` and allows only for a response type `R`.
+///
+/// # Example:
+/// ```
+/// let raw_msg = &subscriber.next().await?;
+/// let msg: ReceivedMessage<RequestConfig, ReplyConfig> =
+///             raw_msg.try_into()?;
+///
+/// msg.respond(ReplyConfig {}).await.unwrap();
+/// // or we can also use the same fn to return an error
+/// msg.respond(Err(Error::Message("failure".into()))).await.unwrap();
+/// ```
+pub struct ReceivedMessage<'a, S, R> {
+    request: SendPayload<S>,
+    bus_message: &'a nats::asynk::Message,
+    reply_type: PhantomData<R>,
+}
+
+impl<'a, S, R> ReceivedMessage<'a, S, R>
+where
+    for<'de> S: Deserialize<'de> + 'a + Debug + Clone + MessageIdTrait,
+    R: Serialize,
+{
+    /// Get a copy of the actual payload data which was received.
+    pub fn payload(&self) -> S {
+        self.request.data.clone()
+    }
+    /// Get a reference of the actual payload data which was received.
+    pub fn payload_as_ref(&self) -> &S {
+        &self.request.data
```

Comment:
Would it to better to use the AsRef trait here?

Reply:
I guess S needs to implement the trait. I've removed it for now. If we use proc-macros we might be able to automatically add that trait and any others to the message bus types.

Reply:
Yeah. this is still harder than it should be to get right.